File: gymates_flutter/lib/pages/training_new/models/training_models.dart
/// 🏋️ 训练模块数据模型

class TrainingPlan {
  final String id;
  final String name;
  final String description;
  final int duration;
  final int caloriesBurned;
  final String difficulty;
  final bool isPublic;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<TrainingExercise> exercises;

  TrainingPlan({
    required this.id,
    required this.name,
    required this.description,
    required this.duration,
    required this.caloriesBurned,
    required this.difficulty,
    required this.isPublic,
    required this.createdAt,
    required this.updatedAt,
    required this.exercises,
  });

  factory TrainingPlan.fromJson(Map<String, dynamic> json) {
    return TrainingPlan(
      id: json['id'].toString(),
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      duration: json['duration'] ?? 0,
      caloriesBurned: json['calories_burned'] ?? 0,
      difficulty: json['difficulty'] ?? 'beginner',
      isPublic: json['is_public'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      exercises: (json['exercises'] as List<dynamic>?)
              ?.map((e) => TrainingExercise.fromJson(e))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'duration': duration,
      'calories_burned': caloriesBurned,
      'difficulty': difficulty,
      'is_public': isPublic,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'exercises': exercises.map((e) => e.toJson()).toList(),
    };
  }
}

class TrainingExercise {
  final String id;
  final String name;
  final int sets;
  final int reps;
  final double? weight;
  final int? duration;
  final int restTime;
  final String? instructions;
  final String? imageUrl;
  final String? muscleGroup;
  final String? difficulty;
  final String? equipment;
  final int order;

  TrainingExercise({
    required this.id,
    required this.name,
    required this.sets,
    required this.reps,
    this.weight,
    this.duration,
    required this.restTime,
    this.instructions,
    this.imageUrl,
    this.muscleGroup,
    this.difficulty,
    this.equipment,
    required this.order,
  });

  factory TrainingExercise.fromJson(Map<String, dynamic> json) {
    return TrainingExercise(
      id: json['id'].toString(),
      name: json['name'] ?? '',
      sets: json['sets'] ?? 0,
      reps: json['reps'] ?? 0,
      weight: json['weight']?.toDouble(),
      duration: json['duration'],
      restTime: json['rest_time'] ?? 0,
      instructions: json['instructions'],
      imageUrl: json['image_url'],
      muscleGroup: json['muscle_group'],
      difficulty: json['difficulty'],
      equipment: json['equipment'],
      order: json['order'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'sets': sets,
      'reps': reps,
      'weight': weight,
      'duration': duration,
      'rest_time': restTime,
      'instructions': instructions,
      'image_url': imageUrl,
      'muscle_group': muscleGroup,
      'difficulty': difficulty,
      'equipment': equipment,
      'order': order,
    };
  }
}

class TodayTrainingPlan {
  final String? planId;
  final List<TodayExercise> exercises;
  final bool isAIRecommended;
  final DateTime? createdAt;

  TodayTrainingPlan({
    this.planId,
    required this.exercises,
    this.isAIRecommended = false,
    this.createdAt,
  });

  factory TodayTrainingPlan.fromJson(Map<String, dynamic> json) {
    return TodayTrainingPlan(
      planId: json['plan_id'],
      exercises: (json['exercises'] as List<dynamic>?)
              ?.map((e) => TodayExercise.fromJson(e))
              .toList() ??
          [],
      isAIRecommended: json['is_ai_recommended'] ?? false,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'plan_id': planId,
      'exercises': exercises.map((e) => e.toJson()).toList(),
      'is_ai_recommended': isAIRecommended,
      'created_at': createdAt?.toIso8601String(),
    };
  }
}

class TodayExercise {
  final String id;
  final String name;
  final int sets;
  final int reps;
  final int restSeconds;
  final String muscleGroup;
  final bool isCompleted;

  TodayExercise({
    required this.id,
    required this.name,
    required this.sets,
    required this.reps,
    required this.restSeconds,
    required this.muscleGroup,
    this.isCompleted = false,
  });

  factory TodayExercise.fromJson(Map<String, dynamic> json) {
    return TodayExercise(
      id: json['id'].toString(),
      name: json['name'] ?? '',
      sets: json['sets'] ?? 0,
      reps: json['reps'] ?? 0,
      restSeconds: json['rest_seconds'] ?? json['rest_time'] ?? 60,
      muscleGroup: json['muscle_group'] ?? '',
      isCompleted: json['is_completed'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'sets': sets,
      'reps': reps,
      'rest_seconds': restSeconds,
      'muscle_group': muscleGroup,
      'is_completed': isCompleted,
    };
  }

  TodayExercise copyWith({
    String? id,
    String? name,
    int? sets,
    int? reps,
    int? restSeconds,
    String? muscleGroup,
    bool? isCompleted,
  }) {
    return TodayExercise(
      id: id ?? this.id,
      name: name ?? this.name,
      sets: sets ?? this.sets,
      reps: reps ?? this.reps,
      restSeconds: restSeconds ?? this.restSeconds,
      muscleGroup: muscleGroup ?? this.muscleGroup,
      isCompleted: isCompleted ?? this.isCompleted,
    );
  }
}

class TrainingSession {
  final String id;
  final String userId;
  final String trainingPlanId;
  final String status; // ongoing, completed, paused
  final int progress; // 0-100
  final DateTime startTime;
  final DateTime? endTime;
  final int? duration;
  final int? caloriesBurned;

  TrainingSession({
    required this.id,
    required this.userId,
    required this.trainingPlanId,
    required this.status,
    required this.progress,
    required this.startTime,
    this.endTime,
    this.duration,
    this.caloriesBurned,
  });

  factory TrainingSession.fromJson(Map<String, dynamic> json) {
    return TrainingSession(
      id: json['id'].toString(),
      userId: json['user_id'].toString(),
      trainingPlanId: json['training_plan_id'].toString(),
      status: json['status'] ?? 'ongoing',
      progress: json['progress'] ?? 0,
      startTime: DateTime.parse(json['start_time']),
      endTime: json['end_time'] != null
          ? DateTime.parse(json['end_time'])
          : null,
      duration: json['duration'],
      caloriesBurned: json['calories_burned'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'training_plan_id': trainingPlanId,
      'status': status,
      'progress': progress,
      'start_time': startTime.toIso8601String(),
      'end_time': endTime?.toIso8601String(),
      'duration': duration,
      'calories_burned': caloriesBurned,
    };
  }
}

class TrainingHistory {
  final String id;
  final String trainingPlanName;
  final DateTime date;
  final int duration;
  final int caloriesBurned;
  final int completedExercises;
  final int totalExercises;
  final double completionRate;

  TrainingHistory({
    required this.id,
    required this.trainingPlanName,
    required this.date,
    required this.duration,
    required this.caloriesBurned,
    required this.completedExercises,
    required this.totalExercises,
    required this.completionRate,
  });

  factory TrainingHistory.fromJson(Map<String, dynamic> json) {
    final completed = json['completed_exercises'] ?? 0;
    final total = json['total_exercises'] ?? 1;
    return TrainingHistory(
      id: json['id'].toString(),
      trainingPlanName: json['training_plan_name'] ?? '',
      date: DateTime.parse(json['date']),
      duration: json['duration'] ?? 0,
      caloriesBurned: json['calories_burned'] ?? 0,
      completedExercises: completed,
      totalExercises: total,
      completionRate: total > 0 ? completed / total * 100 : 0,
    );
  }
}

class AIRecommendation {
  final String id;
  final List<TodayExercise> exercises;
  final String reason;
  final DateTime createdAt;
  final String goal;
  final String level;

  AIRecommendation({
    required this.id,
    required this.exercises,
    required this.reason,
    required this.createdAt,
    required this.goal,
    required this.level,
  });

  factory AIRecommendation.fromJson(Map<String, dynamic> json) {
    return AIRecommendation(
      id: json['id'].toString(),
      exercises: (json['exercises'] as List<dynamic>?)
              ?.map((e) => TodayExercise.fromJson(e))
              .toList() ??
          [],
      reason: json['reason'] ?? '',
      createdAt: DateTime.parse(json['created_at']),
      goal: json['goal'] ?? '',
      level: json['level'] ?? '',
    );
  }
}

File: backend/models/training_models.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// UserExerciseFavorite 已在 exercise_library.go 中定义

// TrainingPlanV2 增强的训练计划模型
type TrainingPlanV2 struct {
	ID                uint             `json:"id" gorm:"primaryKey"`
	UserID            uint             `json:"user_id" gorm:"not null;index"`
	User              User             `json:"user" gorm:"foreignKey:UserID"`
	Name              string           `json:"name" gorm:"size:100;not null"`
	Description       string           `json:"description" gorm:"type:text"`
	Difficulty        string           `json:"difficulty" gorm:"size:20;default:'beginner'"`
	Goal              string           `json:"goal" gorm:"size:50"`
	IsAIGenerated     bool             `json:"is_ai_generated" gorm:"default:false"`
	IsPublic          bool             `json:"is_public" gorm:"default:false"`
	ImageURL          string           `json:"image_url" gorm:"size:255"`
	EstimatedDuration int              `json:"estimated_duration"`
	EstimatedCalories int              `json:"estimated_calories"`
	Exercises         []PlanExerciseV2 `json:"exercises" gorm:"foreignKey:TrainingPlanID"`
	CreatedAt         time.Time        `json:"created_at"`
	UpdatedAt         time.Time        `json:"updated_at"`
	DeletedAt         gorm.DeletedAt   `json:"-" gorm:"index"`
}

// PlanExerciseV2 训练计划中的运动项
type PlanExerciseV2 struct {
	ID             uint            `json:"id" gorm:"primaryKey"`
	TrainingPlanID uint            `json:"training_plan_id" gorm:"not null;index"`
	ExerciseID     uint            `json:"exercise_id" gorm:"not null"`
	Exercise       ExerciseLibrary `json:"exercise" gorm:"foreignKey:ExerciseID"`
	Sets           int             `json:"sets" gorm:"not null"`
	Reps           int             `json:"reps" gorm:"not null"`
	Weight         float64         `json:"weight"`
	Duration       int             `json:"duration"`
	RestTime       int             `json:"rest_time" gorm:"default:60"`
	Order          int             `json:"order" gorm:"not null"`
	Notes          string          `json:"notes" gorm:"type:text"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
}

// TodayWorkout 今日训练
type TodayWorkout struct {
	ID        uint                `json:"id" gorm:"primaryKey"`
	UserID    uint                `json:"user_id" gorm:"not null;index"`
	User      User                `json:"user" gorm:"foreignKey:UserID"`
	PlanID    *uint               `json:"plan_id"`
	Plan      *TrainingPlanV2     `json:"plan" gorm:"foreignKey:PlanID"`
	Date      time.Time           `json:"date" gorm:"not null;index"`
	Status    string              `json:"status" gorm:"size:20;default:'not_started'"` // not_started, in_progress, completed
	SessionID *uint               `json:"session_id"`
	Session   *WorkoutSessionV2   `json:"session" gorm:"foreignKey:SessionID"`
	Exercises []WorkoutExerciseV2 `json:"exercises" gorm:"foreignKey:WorkoutID"`
	CreatedAt time.Time           `json:"created_at"`
	UpdatedAt time.Time           `json:"updated_at"`
}

// WorkoutSessionV2 训练会话 V2
type WorkoutSessionV2 struct {
	ID             uint            `json:"id" gorm:"primaryKey"`
	UserID         uint            `json:"user_id" gorm:"not null;index"`
	User           User            `json:"user" gorm:"foreignKey:UserID"`
	PlanID         *uint           `json:"plan_id"`
	Plan           *TrainingPlanV2 `json:"plan" gorm:"foreignKey:PlanID"`
	IsAIWorkout    bool            `json:"is_ai_workout" gorm:"default:false"`
	StartTime      time.Time       `json:"start_time" gorm:"not null"`
	EndTime        *time.Time      `json:"end_time"`
	Status         string          `json:"status" gorm:"size:20;default:'ongoing'"` // ongoing, completed, abandoned
	Duration       int             `json:"duration"`                                // minutes
	CaloriesBurned int             `json:"calories_burned"`
	Progress       int             `json:"progress" gorm:"default:0"` // 0-100
	Notes          string          `json:"notes" gorm:"type:text"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
	DeletedAt      gorm.DeletedAt  `json:"-" gorm:"index"`
}

// WorkoutExerciseV2 训练中的运动项
type WorkoutExerciseV2 struct {
	ID            uint            `json:"id" gorm:"primaryKey"`
	WorkoutID     uint            `json:"workout_id" gorm:"not null;index"`
	Workout       TodayWorkout    `json:"workout" gorm:"foreignKey:WorkoutID"`
	ExerciseID    uint            `json:"exercise_id" gorm:"not null"`
	Exercise      ExerciseLibrary `json:"exercise" gorm:"foreignKey:ExerciseID"`
	TotalSets     int             `json:"total_sets" gorm:"not null"`
	CompletedSets int             `json:"completed_sets" gorm:"default:0"`
	TargetReps    int             `json:"target_reps"`
	TargetWeight  float64         `json:"target_weight"`
	SetRecords    []SetRecord     `json:"set_records" gorm:"foreignKey:WorkoutExerciseID"`
	Order         int             `json:"order" gorm:"not null"`
	CreatedAt     time.Time       `json:"created_at"`
	UpdatedAt     time.Time       `json:"updated_at"`
}

// SetRecord 单组记录
type SetRecord struct {
	ID                uint      `json:"id" gorm:"primaryKey"`
	WorkoutExerciseID uint      `json:"workout_exercise_id" gorm:"not null;index"`
	SetNumber         int       `json:"set_number" gorm:"not null"`
	Reps              int       `json:"reps" gorm:"not null"`
	Weight            float64   `json:"weight"`
	Duration          int       `json:"duration"`
	CompletedAt       time.Time `json:"completed_at" gorm:"not null"`
	Notes             string    `json:"notes" gorm:"type:text"`
	CreatedAt         time.Time `json:"created_at"`
}

// TrainingHistory 训练历史记录
type TrainingHistory struct {
	ID                 uint             `json:"id" gorm:"primaryKey"`
	UserID             uint             `json:"user_id" gorm:"not null;index"`
	User               User             `json:"user" gorm:"foreignKey:UserID"`
	SessionID          uint             `json:"session_id" gorm:"not null"`
	Session            WorkoutSessionV2 `json:"session" gorm:"foreignKey:SessionID"`
	PlanID             *uint            `json:"plan_id"`
	PlanName           string           `json:"plan_name" gorm:"size:100"`
	Date               time.Time        `json:"date" gorm:"not null;index"`
	Duration           int              `json:"duration"` // minutes
	CaloriesBurned     int              `json:"calories_burned"`
	CompletedExercises int              `json:"completed_exercises"`
	TotalExercises     int              `json:"total_exercises"`
	CompletionRate     float64          `json:"completion_rate"`
	IsAIWorkout        bool             `json:"is_ai_workout" gorm:"default:false"`
	Notes              string           `json:"notes" gorm:"type:text"`
	CreatedAt          time.Time        `json:"created_at"`
	UpdatedAt          time.Time        `json:"updated_at"`
}

// UserTrainingStats 用户训练统计
type UserTrainingStats struct {
	ID                   uint       `json:"id" gorm:"primaryKey"`
	UserID               uint       `json:"user_id" gorm:"uniqueIndex;not null"`
	User                 User       `json:"user" gorm:"foreignKey:UserID"`
	TotalWorkouts        int        `json:"total_workouts" gorm:"default:0"`
	TotalMinutes         int        `json:"total_minutes" gorm:"default:0"`
	TotalCaloriesBurned  int        `json:"total_calories_burned" gorm:"default:0"`
	CurrentStreak        int        `json:"current_streak" gorm:"default:0"`
	LongestStreak        int        `json:"longest_streak" gorm:"default:0"`
	LastWorkoutDate      *time.Time `json:"last_workout_date"`
	MuscleGroupFrequency string     `json:"muscle_group_frequency" gorm:"type:text"` // JSON格式
	RecentActivities     string     `json:"recent_activities" gorm:"type:text"`      // JSON格式
	AverageIntensity     float64    `json:"average_intensity"`
	UpdatedAt            time.Time  `json:"updated_at"`
}

// AIRecommendation AI训练推荐
type AIRecommendation struct {
	ID        uint            `json:"id" gorm:"primaryKey"`
	UserID    uint            `json:"user_id" gorm:"not null;index"`
	User      User            `json:"user" gorm:"foreignKey:UserID"`
	PlanID    *uint           `json:"plan_id"`
	Plan      *TrainingPlanV2 `json:"plan" gorm:"foreignKey:PlanID"`
	Reason    string          `json:"reason" gorm:"type:text"`
	Goal      string          `json:"goal" gorm:"size:50"`
	Level     string          `json:"level" gorm:"size:20"`
	Exercises string          `json:"exercises" gorm:"type:text"` // JSON格式的运动列表
	IsApplied bool            `json:"is_applied" gorm:"default:false"`
	CreatedAt time.Time       `json:"created_at"`
	DeletedAt gorm.DeletedAt  `json:"-" gorm:"index"`
}

// TableName 指定表名
// TableName 方法已在 exercise_library.go 中定义

func (TrainingPlanV2) TableName() string {
	return "training_plans_v2"
}

func (PlanExerciseV2) TableName() string {
	return "plan_exercises_v2"
}

func (TodayWorkout) TableName() string {
	return "today_workouts"
}

func (WorkoutSessionV2) TableName() string {
	return "workout_sessions_v2"
}

func (WorkoutExerciseV2) TableName() string {
	return "workout_exercises_v2"
}

func (SetRecord) TableName() string {
	return "set_records"
}

func (TrainingHistory) TableName() string {
	return "training_histories"
}

func (UserTrainingStats) TableName() string {
	return "user_training_stats"
}

func (AIRecommendation) TableName() string {
	return "ai_recommendations"
}
